How can I get TypeScript `baseUrl` to work?

The `v1.18.0` [release notes](https://github.com/microsoft/playwright/releases/tag/v1.18.0) mention support for `baseUrl` in `tsconfig.json`. I'm not able to make it work.

Here is a minimal example: https://github.com/joakimgunst/playwright-debug

I have `baseUrl` set to `./src`, and import `src/titles.ts` as `from "titles"` in `e2e/example.spec.ts`.

When I run `npm run test` in the example repo, I get `Error: Cannot find module 'titles'`.

TypeScript does not show any errors when running `npm run tsc`.

If I change the import to `from "../src/titles"`, everything works as expected.

Also, thank you for creating this great library!

--- a/packages/playwright-core/src/cli/cli.ts
+++ b/packages/playwright-core/src/cli/cli.ts
@@ -15,10 +15,12 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+
 import { fork } from 'child_process';
+import url from 'url';
 
 if (process.env.PW_EXPERIMENTAL_TS_ESM) {
-  const NODE_OPTIONS = (process.env.NODE_OPTIONS || '') + ` --experimental-loader=${require.resolve('@playwright/test/lib/experimentalLoader')}`;
+  const NODE_OPTIONS = (process.env.NODE_OPTIONS || '') + ` --experimental-loader=${url.pathToFileURL(require.resolve('@playwright/test/lib/experimentalLoader'))}`;
   const innerProcess = fork(require.resolve('./innerCli'), process.argv.slice(2), {
     env: { ...process.env, NODE_OPTIONS }
   });

--- a/packages/playwright-test/package.json
+++ b/packages/playwright-test/package.json
@@ -45,7 +45,6 @@
     "@babel/plugin-syntax-optional-catch-binding": "7.8.3",
     "@babel/plugin-transform-modules-commonjs": "7.16.8",
     "@babel/preset-typescript": "7.16.7",
-    "babel-plugin-module-resolver": "4.1.0",
     "colors": "1.4.0",
     "commander": "8.3.0",
     "debug": "4.3.3",

--- a/packages/playwright-test/src/experimentalLoader.ts
+++ b/packages/playwright-test/src/experimentalLoader.ts
@@ -15,28 +15,27 @@
  */
 
 import fs from 'fs';
-import { transformHook } from './transform';
+import url from 'url';
+import { transformHook, resolveHook } from './transform';
 
 async function resolve(specifier: string, context: { parentURL: string }, defaultResolve: any) {
-  if (specifier.endsWith('.js') || specifier.endsWith('.ts') || specifier.endsWith('.mjs'))
-    return defaultResolve(specifier, context, defaultResolve);
-  let url = new URL(specifier, context.parentURL).toString();
-  url = url.substring('file://'.length);
-  for (const extension of ['.ts', '.js', '.tsx', '.jsx']) {
-    if (fs.existsSync(url + extension))
-      return defaultResolve(specifier + extension, context, defaultResolve);
+  if (context.parentURL && context.parentURL.startsWith('file://')) {
+    const filename = url.fileURLToPath(context.parentURL);
+    const resolved = resolveHook(filename, specifier);
+    if (resolved !== undefined)
+      specifier = url.pathToFileURL(resolved).toString();
   }
   return defaultResolve(specifier, context, defaultResolve);
 }
 
-async function load(url: string, context: any, defaultLoad: any) {
-  if (url.endsWith('.ts') || url.endsWith('.tsx')) {
-    const filename = url.substring('file://'.length);
+async function load(moduleUrl: string, context: any, defaultLoad: any) {
+  if (moduleUrl.startsWith('file://') && (moduleUrl.endsWith('.ts') || moduleUrl.endsWith('.tsx'))) {
+    const filename = url.fileURLToPath(moduleUrl);
     const code = fs.readFileSync(filename, 'utf-8');
     const source = transformHook(code, filename, true);
     return { format: 'module', source };
   }
-  return defaultLoad(url, context, defaultLoad);
+  return defaultLoad(moduleUrl, context, defaultLoad);
 }
 
 module.exports = { resolve, load };

--- a/packages/playwright-test/src/transform.ts
+++ b/packages/playwright-test/src/transform.ts
@@ -23,16 +23,15 @@ import * as sourceMapSupport from 'source-map-support';
 import * as url from 'url';
 import type { Location } from './types';
 import { tsConfigLoader, TsConfigLoaderResult } from './third_party/tsconfig-loader';
+import Module from 'module';
 
-const version = 6;
+const version = 7;
 const cacheDir = process.env.PWTEST_CACHE_DIR || path.join(os.tmpdir(), 'playwright-transform-cache');
 const sourceMaps: Map<string, string> = new Map();
 
 type ParsedTsConfigData = {
-  absoluteBaseUrl: string,
-  singlePath: { [key: string]: string },
-  hash: string,
-  alias: { [key: string]: string | ((s: string[]) => string) },
+  absoluteBaseUrl: string;
+  paths: { key: string, values: string[] }[];
 };
 const cachedTSConfigs = new Map<string, ParsedTsConfigData | undefined>();
 
@@ -55,9 +54,8 @@ sourceMapSupport.install({
   }
 });
 
-function calculateCachePath(tsconfigData: ParsedTsConfigData | undefined, content: string, filePath: string): string {
+function calculateCachePath(content: string, filePath: string): string {
   const hash = crypto.createHash('sha1')
-      .update(tsconfigData?.hash || '')
       .update(process.env.PW_TEST_SOURCE_TRANSFORM || '')
       .update(process.env.PW_EXPERIMENTAL_TS_ESM ? 'esm' : 'no_esm')
       .update(content)
@@ -69,43 +67,12 @@ function calculateCachePath(tsconfigData: ParsedTsConfigData | undefined, conten
 }
 
 function validateTsConfig(tsconfig: TsConfigLoaderResult): ParsedTsConfigData | undefined {
-  if (!tsconfig.tsConfigPath || !tsconfig.paths || !tsconfig.baseUrl)
+  if (!tsconfig.tsConfigPath || !tsconfig.baseUrl)
     return;
-
-  const paths = tsconfig.paths;
-  // Path that only contains "*", ".", "/" and "\" is too ambiguous.
-  const ambiguousPath = Object.keys(paths).find(key => key.match(/^[*./\\]+$/));
-  if (ambiguousPath)
-    return;
-  const multiplePath = Object.keys(paths).find(key => paths[key].length > 1);
-  if (multiplePath)
-    return;
-  // Only leave a single path mapping.
-  const singlePath = Object.fromEntries(Object.entries(paths).map(([key, values]) => ([key, values[0]])));
   // Make 'baseUrl' absolute, because it is relative to the tsconfig.json, not to cwd.
   const absoluteBaseUrl = path.resolve(path.dirname(tsconfig.tsConfigPath), tsconfig.baseUrl);
-  const hash = JSON.stringify({ absoluteBaseUrl, singlePath });
-
-  const alias: ParsedTsConfigData['alias'] = {};
-  for (const [key, value] of Object.entries(singlePath)) {
-    const regexKey = '^' + key.replace('*', '.*');
-    alias[regexKey] = ([name]) => {
-      let relative: string;
-      if (key.endsWith('/*'))
-        relative = value.substring(0, value.length - 1) + name.substring(key.length - 1);
-      else
-        relative = value;
-      relative = relative.replace(/\//g, path.sep);
-      return path.resolve(absoluteBaseUrl, relative);
-    };
-  }
-
-  return {
-    absoluteBaseUrl,
-    singlePath,
-    hash,
-    alias,
-  };
+  const paths = tsconfig.paths || { '*': ['*'] };
+  return { absoluteBaseUrl, paths: Object.entries(paths).map(([key, values]) => ({ key, values })) };
 }
 
 function loadAndValidateTsconfigForFile(file: string): ParsedTsConfigData | undefined {
@@ -123,6 +90,35 @@ function loadAndValidateTsconfigForFile(file: string): ParsedTsConfigData | unde
 const pathSeparator = process.platform === 'win32' ? ';' : ':';
 const scriptPreprocessor = process.env.PW_TEST_SOURCE_TRANSFORM ?
   require(process.env.PW_TEST_SOURCE_TRANSFORM) : undefined;
+const builtins = new Set(Module.builtinModules);
+
+export function resolveHook(filename: string, specifier: string): string | undefined {
+  if (builtins.has(specifier))
+    return;
+  const isTypeScript = filename.endsWith('.ts') || filename.endsWith('.tsx');
+  if (!isTypeScript)
+    return;
+  const tsconfig = loadAndValidateTsconfigForFile(filename);
+  if (!tsconfig)
+    return;
+  for (const { key, values } of tsconfig.paths) {
+    const keyHasStar = key[key.length - 1] === '*';
+    const matches = specifier.startsWith(keyHasStar ? key.substring(0, key.length - 1) : key);
+    if (!matches)
+      continue;
+    for (const value of values) {
+      const valueHasStar = value[value.length - 1] === '*';
+      let candidate = valueHasStar ? value.substring(0, value.length - 1) : value;
+      if (valueHasStar && keyHasStar)
+        candidate += specifier.substring(key.length - 1);
+      candidate = path.resolve(tsconfig.absoluteBaseUrl, candidate.replace(/\//g, path.sep));
+      for (const ext of ['', '.js', '.ts', '.mjs', '.cjs', '.jsx', '.tsx']) {
+        if (fs.existsSync(candidate + ext))
+          return candidate;
+      }
+    }
+  }
+}
 
 export function transformHook(code: string, filename: string, isModule = false): string {
   if (isComponentImport(filename))
@@ -138,8 +134,7 @@ export function transformHook(code: string, filename: string, isModule = false):
   if (!isTypeScript && !hasPreprocessor)
     return code;
 
-  const tsconfigData = isTypeScript ? loadAndValidateTsconfigForFile(filename) : undefined;
-  const cachePath = calculateCachePath(tsconfigData, code, filename);
+  const cachePath = calculateCachePath(code, filename);
   const codePath = cachePath + '.js';
   const sourceMapPath = cachePath + '.map';
   sourceMaps.set(filename, sourceMapPath);
@@ -165,16 +160,6 @@ export function transformHook(code: string, filename: string, isModule = false):
         [require.resolve('@babel/plugin-syntax-object-rest-spread')],
         [require.resolve('@babel/plugin-proposal-export-namespace-from')]
     );
-
-    if (tsconfigData) {
-      plugins.push([require.resolve('babel-plugin-module-resolver'), {
-        root: ['./'],
-        alias: tsconfigData.alias,
-        // Silences warning 'Could not resovle ...' that we trigger because we resolve
-        // into 'foo/bar', and not 'foo/bar.ts'.
-        loglevel: 'silent',
-      }]);
-    }
 
     if (!isModule) {
       plugins.push([require.resolve('@babel/plugin-transform-modules-commonjs')]);
@@ -217,12 +202,30 @@ export function transformHook(code: string, filename: string, isModule = false):
 }
 
 export function installTransform(): () => void {
-  const exts = ['.ts', '.tsx'];
+  let reverted = false;
 
+  const originalResolveFilename = (Module as any)._resolveFilename;
+  function resolveFilename(this: any, specifier: string, parent: Module, ...rest: any[]) {
+    if (!reverted && parent) {
+      const resolved = resolveHook(parent.filename, specifier);
+      if (resolved !== undefined)
+        specifier = resolved;
+    }
+    return originalResolveFilename.call(this, specifier, parent, ...rest);
+  }
+  (Module as any)._resolveFilename = resolveFilename;
+
+  const exts = ['.ts', '.tsx'];
   // When script preprocessor is engaged, we transpile JS as well.
   if (scriptPreprocessor)
     exts.push('.js', '.mjs');
-  return pirates.addHook((code: string, filename: string) => transformHook(code, filename), { exts });
+  const revertPirates = pirates.addHook((code: string, filename: string) => transformHook(code, filename), { exts });
+
+  return () => {
+    reverted = true;
+    (Module as any)._resolveFilename = originalResolveFilename;
+    revertPirates();
+  };
 }
 
 export function wrapFunctionWithLocation<A extends any[], R>(func: (location: Location, ...args: A) => R): (...args: A) => R {

--- a/tests/playwright-test/resolver.spec.ts
+++ b/tests/playwright-test/resolver.spec.ts
@@ -34,7 +34,7 @@ test('should respect path resolver', async ({ runInlineTest }) => {
         "paths": {
           "util/*": ["./foo/bar/util/*"],
           "util2/*": ["./foo/bar/util/*"],
-          "util3": ["./foo/bar/util/b"],
+          "util3": ["./does-not-exist", "./foo/bar/util/b"],
         },
       },
     }`,
@@ -136,72 +136,32 @@ test('should respect baseurl', async ({ runInlineTest }) => {
   expect(result.exitCode).toBe(0);
 });
 
-test('should ignore for baseurl w/o paths', async ({ runInlineTest }) => {
+test('should respect baseurl w/o paths', async ({ runInlineTest }) => {
   const result = await runInlineTest({
-    'tsconfig.json': `{
+    'foo/bar/util/b.ts': `
+      export const foo = 42;
+    `,
+    'dir2/tsconfig.json': `{
       "compilerOptions": {
         "target": "ES2019",
         "module": "commonjs",
         "lib": ["esnext", "dom", "DOM.Iterable"],
-        "baseUrl": "./"
+        "baseUrl": "..",
       },
     }`,
-    'a.test.ts': `
-      import { foo } from 'foo/b';
-    `,
-    'foo/b.ts': `
-      export const foo = 1;
+    'dir2/inner.spec.ts': `
+      // This import should pick up ../foo/bar/util/b due to baseUrl.
+      import { foo } from 'foo/bar/util/b';
+      const { test } = pwt;
+      test('test', ({}, testInfo) => {
+        expect(foo).toBe(42);
+      });
     `,
   });
 
-  expect(result.exitCode).toBe(1);
-  expect(result.output).toContain(`Cannot find module 'foo/b'`);
-});
-
-test('should ignore tsconfig with ambiguous paths', async ({ runInlineTest }) => {
-  const result = await runInlineTest({
-    'tsconfig.json': `{
-      "compilerOptions": {
-        "target": "ES2019",
-        "module": "commonjs",
-        "lib": ["esnext", "dom", "DOM.Iterable"],
-        "baseUrl": "./",
-        "paths": { "*/*": ["*/*"] }
-      },
-    }`,
-    'a.test.ts': `
-      import { foo } from 'foo/b';
-    `,
-    'foo/b.ts': `
-      export const foo = 1;
-    `,
-  });
-
-  expect(result.exitCode).toBe(1);
-  expect(result.output).toContain(`Cannot find module 'foo/b'`);
-});
-
-test('should ignore tsconfig with multi-value paths', async ({ runInlineTest }) => {
-  const result = await runInlineTest({
-    'tsconfig.json': `{
-      "compilerOptions": {
-        "target": "ES2019",
-        "module": "commonjs",
-        "lib": ["esnext", "dom", "DOM.Iterable"],
-        "baseUrl": "./",
-        "paths": { "foo/*": ["./foo/*", "./bar/*"] }
-      },
-    }`,
-    'a.test.ts': `
-      import { foo } from 'foo/b';
-    `,
-    'foo/b.ts': `
-      export const foo = 1;
-    `,
-  });
-
-  expect(result.exitCode).toBe(1);
-  expect(result.output).toContain(`Cannot find module 'foo/b'`);
+  expect(result.exitCode).toBe(0);
+  expect(result.passed).toBe(1);
+  expect(result.output).not.toContain(`Could not`);
 });
 
 test('should respect path resolver in experimental mode', async ({ runInlineTest }) => {

